### PR TITLE
Improve safe transmute error reporting

### PR DESCRIPTION
**Pull request by bryangarza**
This patch updates the error reporting when Safe Transmute is not possible between 2 types by including the reason.

Also, fix some small bugs that occur when computing the `Answer` for transmutability.

### CODE REVIEW

**Comment by bryangarza** [comment]
cc @jswrenn

**Comment by bryangarza** [comment]
I think I missed some tests that need to be blessed, I’ll do it tomorrow

File: compiler/rustc_transmute/Cargo.toml
Comment (jswrenn) [inline compiler/rustc_transmute/Cargo.toml:17]: I don't think this can be optional. To check, `cd` into `rustc_transmute` and run `cargo check`.

File: tests/ui/transmutability/enums/repr/should_require_well_defined_layout.stderr
Comment (jswrenn) [inline tests/ui/transmutability/enums/repr/should_require_well_defined_layout.stderr:8]: Instead of:
```
`void::repr_rust` cannot be safely transmuted into `()` in the defining scope of `assert::Context`.
Reason: SrcIsUnspecified
```
I'd suggest:
```
`void::repr_rust` cannot be safely transmuted into `()` in the defining scope of `assert::Context` because `void::repr_rust` does not have a well-specified layout.
```

File: tests/ui/transmutability/enums/should_pad_variants.stderr
Comment (jswrenn) [inline tests/ui/transmutability/enums/should_pad_variants.stderr:8]: I'd suggest:
```
`Src` cannot be safely transmuted into `Dst` in the defining scope of `should_pad_variants::Context` because the size of `Src` is smaller than the size of `Dst` (YYY bytes).
```

File: tests/ui/transmutability/enums/should_respect_endianness.stderr
Comment (jswrenn) [inline tests/ui/transmutability/enums/should_respect_endianness.stderr:8]: I'd suggest:
```
`Src` cannot be safely transmuted into `Unexpected` in the defining scope of `assert::Context` because at least one value of `Src` isn't a bit-valid value of `Unexpected`.
```

Eventually, we'll want to add more information here about why (namely, by giving an example value of the source type that's not a valid instance of the destination type), but we can punt that to future work.

File: tests/ui/transmutability/visibility/should_reject_if_dst_has_private_field.stderr
Comment (jswrenn) [inline tests/ui/transmutability/visibility/should_reject_if_dst_has_private_field.stderr:8]: I'd suggest:
```
`Src` cannot be safely transmuted into `Dst` in the defining scope of `test::Context` because `Dst` is or contains a type or field that is not visible in that scope.
```

**Review by jswrenn — CHANGES_REQUESTED** [review]

File: tests/ui/transmutability/enums/repr/should_require_well_defined_layout.stderr
Comment (bryangarza) [reply tests/ui/transmutability/enums/repr/should_require_well_defined_layout.stderr:8]: Is it possible to combine them somehow? The first line comes from the on_unimplemented macro and the other line comes from the actual error.

File: compiler/rustc_transmute_types/Cargo.toml
Comment (compiler-errors) [inline compiler/rustc_transmute_types/Cargo.toml:1]: I don't think we really need to put this into a separate crate. Most type definitions like this just live in rustc_middle.

File: tests/ui/transmutability/enums/repr/should_require_well_defined_layout.stderr
Comment (compiler-errors) [reply tests/ui/transmutability/enums/repr/should_require_well_defined_layout.stderr:8]: It's probably better to just remove the `rustc_on_unimplemented` from the macro and handle formatting the error message for the transmute trait manually.

**Comment by compiler-errors** [comment]
use `@rustbot ready` to mark this as ready to review

@rustbot author

File: compiler/rustc_transmute/Cargo.toml
Comment (bryangarza) [reply compiler/rustc_transmute/Cargo.toml]: thanks. In my latest commit I fixed the build (had to gate some more code behind the rustc flag because we depend on rustc_middle more)

File: compiler/rustc_middle/src/traits/mod.rs
Comment (bryangarza) [reply compiler/rustc_middle/src/traits/mod.rs]: Thanks, I created a new variant (`rustc_transmute` does a lot of work, so don't want to run it twice)

File: compiler/rustc_middle/src/traits/mod.rs
Comment (bryangarza) [inline compiler/rustc_middle/src/traits/mod.rs:595]: I named the variant `NotTransmutable` because the other variants also use `Not`, and I named the reason type `NonTransmutabilityReason` because `NotTransmutableReason` reads kind of weird IMO, but I can rename them to have the same prefix if we want.

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:620]: I would probably just handle this branch separately.

File: compiler/rustc_trait_selection/src/traits/select/confirmation.rs
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/confirmation.rs:308]: Do you need to return these types? I think you can extract them from the `FulfillmentError` during error reporting?

**Comment by bryangarza** [comment]
FYI
- easier to review if you hide whitespace changes
- All the changes in `rustc_transmute` are just moving some definitions to be under `mod rustc` -- no actual code changes there

File: tests/ui/transmutability/arrays/should_require_well_defined_layout.stderr
Comment (compiler-errors) [inline tests/ui/transmutability/arrays/should_require_well_defined_layout.stderr:5]: This should probably be reverted back to what it used to be.

File: compiler/rustc_trait_selection/src/traits/select/confirmation.rs
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:308]: Also, does this even need to be a separate `SelectionError`? I think you can just check if the [`obligation`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_infer/traits/struct.FulfillmentError.html#structfield.obligation) in the `FulfillmentError` is the transmutability trait, and extract the info again?

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:620]: Actually, see comment below. I'm not convinced this needs a separate `SelectionError`.

File: tests/ui/transmutability/arrays/should_require_well_defined_layout.stderr
Comment (compiler-errors) [reply tests/ui/transmutability/arrays/should_require_well_defined_layout.stderr:5]: Or at least made better, this current rendering is pretty rough.

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
Comment (bryangarza) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:620]: I had it separate but then I realized I would have to add a bunch of the code from this one in order to create the full error (for example, including the code block and pointing to say "required by a bound..."...

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:620]: Hm, this function is getting somewhat out of control. I wonder if you could factor it out in some way to make it more manageable...

Anyways, if you go ahead with my advice below and just use `Unimplemented` then it's kind of moot, but really this could use some investment in clean-up if you're up for it.

File: compiler/rustc_trait_selection/src/traits/select/confirmation.rs
Comment (bryangarza) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:308]: is that the case? I don't understand how to get these fields from `PredicateObligation` -- do you have an example?

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:308]: Well you should have access to the trait predicate - https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_trait_selection/traits/error_reporting/mod.rs.html#653

Then from the trait predicate, you'd just be extracting the types from the substitutions just like the transmute trait confirmation code does.

Comment (bryangarza) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:308]: oh, gotcha, yeah that makes more sense now. I missed that `let trait_predicate = ...` line before.. will try it out

Comment (bryangarza) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs]: Okay, I got it to work. I kept a separate `SelectionError` so that I can include the specific `NotTransmutable` error -- I don't think that is possible to have access to without either having the separate variant or calling rustc_transmute again.

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
Comment (bryangarza) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs]: I kept a separate variant, and was able to show the rest of the original error message by only copying over the necessary logic from the `Unspecified` case. But I can put up another PR that refactors the `Unspecified` case so we can reuse pieces and have it not be all in one giant code block.

**Comment by compiler-errors** [comment]
Some more comments. Other than those, I'm pretty happy with the state of this PR, though.

@rustbot author

File: compiler/rustc_transmute/src/layout/tree.rs
Comment (bryangarza) [reply compiler/rustc_transmute/src/layout/tree.rs:198]: I'll add back the `Unknown` case

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
r=me after rebasing

there may be some problems with the error reporting with the new trait solver, since we *will* need to recompute the cannot-transmute-reason on demand in the error reporting code...

you can probably just leave that though, I can figure it out, unless you want to. ping me if you need help.

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs
Comment (bryangarza) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:736]: @compiler-errors Added this so the new trait solver can emit the right error, is this what you had in mind?

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:736]: I think I asked this before, but why do we even need a special selection error kind? You have shown that it's possible to recompute the transmutability reason on the fly here :)

Comment (bryangarza) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:736]: It was just to avoid computing the safe transmute reason twice, but if the new trait solver is going to need it 2x anyways, we could just remove `SelectionError::NotTransmutable` variant

**Comment by compiler-errors** [comment]
Thanks @bryangarza for dealing with many rounds of review.

@bors r+

**Comment by bryangarza** [comment]
@compiler-errors I force pushed a small change after approval (to get `rustc_transmute` to successfully pass `cargo test`) -- is it possible to get this new commit hash to be the one in the queue?

**Comment by compiler-errors** [comment]
Force-pushing automatically un-approves your PR. I will re-approve it when CI is green.

**Pull request merged**
